frontend/lib/src/components/elements/DeckGlJsonChart/useDeckGl.tsx

```diff
@@ -160,6 +235,66 @@ export const useDeckGl = (props: UseDeckGlProps): UseDeckGlShape => {
       }
     }
 
+    if (copy.layers) {
+      const anyLayersHaveSelection = Object.values(
+        data.selection.indices
+      ).some(layer => layer?.length)
+
+      copy.layers = copy.layers.map(layer => {
+        if (!layer || Array.isArray(layer)) {
+          return layer
+        }
+
+        const layerId = `${layer.id || null}`
+        const selectedIndices = data?.selection?.indices?.[layerId] || []
+
+        const fillFunction = LAYER_TYPE_TO_FILL_FUNCTION[layer["@@type"]]
+
+        if (!fillFunction || !Object.hasOwn(layer, fillFunction)) {
+          return layer
+        }
+
+        const clonedLayer = { ...layer }
+        clonedLayer.updateTriggers = {
+          // Tell Deck.gl to recompute the fill color when the selection changes.
+          // Without this, objects in layers will have stale colors when selection changes.
+          // @see https://deck.gl/docs/api-reference/core/layer#updatetriggers
+          [fillFunction]: [
+            ...(clonedLayer.updateTriggers?.[fillFunction] || []),
+            selectedIndices,
+            anyLayersHaveSelection,
+          ],
+        }
+
+        const shouldUseOriginalFillFunction = !anyLayersHaveSelection
+
+        if (shouldUseOriginalFillFunction) {
+          // If we aren't changing the fill color, we don't need to change the fillFunction
+          return clonedLayer
+        }
+
+        const originalFillFunction = layer[fillFunction] as FillFunction
+
+        const selectedColor = parseToRgba(theme.colors.primary)
```

Comment:
It seems like the selected and unselected color is used as some kind of fallback instead of opacity. What are the situations in which this is used?

Reply:
The original usecase was when we were trying different color strategies. At this point it's left there as a fallback in case there are any issues in parsing the color and applying transparency, this guarantees we at least return the original color value, rather than some undefined behavior.